rufs/src/ufs/mod.rs

```diff
@@ -0,0 +1,148 @@
+use std::{
+	ffi::{OsStr, OsString},
+	fs::File,
+	io::{Cursor, Error as IoError, ErrorKind, Read, Result as IoResult, Seek, SeekFrom},
+	mem::size_of,
+	num::NonZeroU64,
+	os::unix::ffi::{OsStrExt, OsStringExt},
+	path::Path,
+};
+
+mod dir;
+mod inode;
+mod symlink;
+mod xattr;
+
+use crate::{
+	blockreader::BlockReader,
+	data::*,
+	decoder::{Config, Decoder},
+};
+
+#[macro_export]
+macro_rules! err {
+	($name:ident) => {
+		IoError::from_raw_os_error(libc::$name)
+	};
+}
+
+macro_rules! iobail {
+	($kind:expr, $($tk:tt)+) => {
+		return Err(IoError::new($kind, format!($($tk)+)))
+	};
+}
+
+#[derive(Debug, Clone)]
+pub struct Info {
+	pub blocks: u64,
+	pub bfree:  u64,
+	pub files:  u64,
+	pub ffree:  u64,
+	pub bsize:  u32,
+	pub fsize:  u32,
+}
+
+pub struct Ufs<R: Read + Seek> {
+	file:       Decoder<BlockReader<R>>,
+	superblock: Superblock,
+}
+
+impl Ufs<File> {
+	pub fn open(path: &Path) -> IoResult<Self> {
+		let file = BlockReader::open(path)?;
+		Self::new(file)
+	}
+}
+
+impl<R: Read + Seek> Ufs<R> {
+	pub fn new(mut file: BlockReader<R>) -> IoResult<Self> {
+		let pos = SBLOCK_UFS2 as u64 + MAGIC_OFFSET;
+		file.seek(SeekFrom::Start(pos))?;
+		let mut magic = [0u8; 4];
+		file.read_exact(&mut magic)?;
+
+		// magic: 0x19 54 01 19
+		#[cfg(not(fuzzing))]
+		let config = match magic {
+			[0x19, 0x01, 0x54, 0x19] => Config::little(),
+			[0x19, 0x54, 0x01, 0x19] => Config::big(),
+			_ => {
+				iobail!(
+					ErrorKind::InvalidInput,
+					"invalid superblock magic number: {magic:?}"
+				)
+			}
+		};
+		// FIXME: Choose based on hash of input or so, to excercise BE as well with introducing non-determinism
+		#[cfg(fuzzing)]
+		let config = Config::little();
+
+		let mut file = Decoder::new(file, config);
+
+		let superblock: Superblock = file.decode_at(SBLOCK_UFS2 as u64)?;
+		#[cfg(not(fuzzing))]
```

Comment:
Why disable this check when fuzzing?

Reply:
So the fuzzer doesn't have to guess a valid superblock, which would lead to a lot of wasted fuzz cases as they'd fail at this stage without reaching crashes/panics deeper in the code. It's common practice to disable these sorts of checks while fuzzing for this reason.

Reply:
The fuzzer doesn't generate totally random input, right?  It starts from a valid disk image and then randomly mutates it?

Reply:
Yes.

Reply:
Given that it starts from a valid disk image, would you really expect to lost very many runs at this stage?  And even if you disable the magic check here, wouldn't a run with a corrupt superblock still likely fail during SuperBlock::check ?

Reply:
This is no longer relevant, as I just always run the checks, even when fuzzing.

Reply:
Uh, what do you mean by "I just always run the checks"?  Do you have a separate fork or something?  Or is there a commit that you forgot to push?  Because the magic check is still disabled during fuzzing.

Reply:
Oh, I forgot to remove this check. I did the match on the magic.